wgsl/index.bs

```diff
@@ -600,28 +642,334 @@ storage_class
+    * In a uniform buffer, the array element alignment is also a multiple of 16.
+* An array's alignment is the same as its element alignment.
+* A matrix with |N| columns is aligned as if it were an array of |N| column vectors.
+* A structure inherits the worst-case alignment of any of its members.
+    * In a uniform buffer, structure member alignment is also a multiple of 16.
+* The members of a structure are laid out in order: earlier members are appear earlier in
+    the buffer
+
+Additionally we define a value's allocation extent, or memory footprint, which determines
+how many memory locations must be reserved to store that value in host-shareable memory.
+Allocation extent is a determining factor of the minimum size of a buffer that can be bound
+to a uniform buffer variable
+or to a storage buffer variable.
+See [[#resource-layout-compatibility]].
+
+Compared to OpenGL:
+* OpenGL `std140` layout satisfies [SHORTNAME] [=uniform buffer layout=].
+* OpenGL `std430` layout satisifes [SHORTNAME] [=storage buffer layout=].
+* A type |S| with tightest offset and stride assignments satisfying [SHORTNAME] [=uniform buffer layout=] also satisfies OpenGL `std140` layout.
+* A type |S| with tightest offset and stride assignments satisfying [SHORTNAME] [=storage buffer layout=] also satisfies OpenGL `std430` layout.
+* OpenGL supports row-major matrices, but [SHORTNAME] does not.
+
+Compared to [[Vulkan1.2ext|Vulkan &sect; 15.6.4 Offset and Stride Assignment]]:
+* Vulkan *standard buffer layout* maps to [SHORTNAME] [=standard buffer layout rules=] with the following qualifications:
+    * Vulkan allows a vector to be aligned to the size of its scalar component, but [SHORTNAME] requires a more constrained alignment.
+    * The Vulkan `scalarBlockLayout` and `uniformBufferStandardLayout` features do not apply to [SHORTNAME].
+    * The Vulkan concept of *scalar alignment* does not correspond to a concept in [SHORTNAME];
+* The Vulkan *base alignment* for a type |S| corresponds to the [SHORTNAME] alignment requirement for |S| in the `storage` storage class: |Align|(|S|,`storage`).
+* The Vulkan *extended alignment* for a type |S| corresponds to the [SHORTNAME] alignment requirement for |S| in the `uniform` storage class: |Align|(|S|,`uniform`).
+* The Vulkan concept of *improperly straddle* is not permitted in [SHORTNAME], because [SHORTNAME] requires vectors to be aligned to their whole size.
+* Vulkan supports non-32-bit scalar types and vector types with non-32-bit components, but [SHORTNAME] does not.
+* Vulkan supports row-major matrices, but [SHORTNAME] does not.
+* Vulkan allows offsets to be non-monotonic, but [SHORTNAME] does not.
+
+#### Internal Layout of Values #### {#internal-value-layout}
+
+This section describes how the internals of a value are placed in the byte locations
+of a buffer, given an assumed placement of the overall value.
+These layouts depend on the value's type, the storage class of the buffer,
```

Comment:
should there be a separate sentence clarifying the behavior when the same type is used for both uniform and storage buffers?

Reply:
Will add that.